### PR TITLE
Cyclebenchmarking client 

### PR DESCRIPTION
**Pull request by paaige**

### CODE REVIEW

File: cirq-superstaq/example-requirements.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
-notebook>=6.5.5
```
Comment (paaige) [inline cirq-superstaq/example-requirements.txt:1]: revert requirements files

File: cirq-superstaq/cirq_superstaq/service.py [hunk 2]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
         )
 
+    def submit_cb(
+        self,
+        target: str,
+        shots: int,
+        circuits: cirq.Circuit,
+        num_channels: int,
+        num_sequences: int,
+        depths: list[int],
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:983]: nit
```suggestion
        depths: Sequence[int],
```

File: cirq-superstaq/cirq_superstaq/service.py [hunk 3]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+            noise_dict["params"] = (
+                (error_prob,) if isinstance(error_prob, numbers.Number) else error_prob
+            )
+        elif isinstance(noise, cirq.NoiseModel):
+            noise_dict["cirq_noise_model"] = cirq.to_json(noise)
+
+        serialized_circuits = css.serialization.serialize_circuits(circuits)
+
+        return self._client.submit_cb(
+            target,
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:1031]:
```suggestion
            self._resolve_target(target),
```
(so this falls back on `default_target`)

File: general-superstaq/general_superstaq/superstaq_client.py [hunk 4]
```diff
@@ -546,6 +546,67 @@ def process_aces(self, job_id: str) -> list[float]:
...
+            "num_sequences": num_sequences,
+            "depths": depths,
+        }
+
+        if method:
+            json_dict["method"] = method
+        if noise:
+            json_dict["noise"] = noise
+
+        return self.post_request("/cycle_benchmarking", json_dict)
```
Comment (richrines1) [inline general-superstaq/general_superstaq/superstaq_client.py:596]: definitely not super important but if feel like it's a little inconsistent to use `"/cycle_benchmarking"` and `"/cb_fetch"` - maybe this should be `"/cb_submit"`?

File: general-superstaq/general_superstaq/superstaq_client.py [hunk 5]
```diff
@@ -546,6 +546,67 @@ def process_aces(self, job_id: str) -> list[float]:
...
         return self.post_request("/aces_fetch", {"job_id": job_id})
 
+    def submit_cb(
+        self,
+        target: str,
+        shots: int,
+        serialized_circuits: dict[str, str],
+        num_channels: int,
+        num_sequences: int,
+        depths: list[int],
```
Comment (richrines1) [inline general-superstaq/general_superstaq/superstaq_client.py:556]:
```suggestion
        depths: Sequence[int],
```

File: cirq-superstaq/cirq_superstaq/service.py [hunk 6]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+            num_channels,
+            num_sequences,
+            depths,
+            method,
+            noise=noise_dict,
+        )
+
+    def process_cb(self, job_id: str) -> dict[str, Any]:
+        """Processes the data from the Cycle Benchmarking protocol.
+        Generates SPAM and decay parameter estimations in addition to the process infidelity.
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:1043]:
```suggestion
        """Processes the data from the Cycle Benchmarking protocol.

        Generates SPAM and decay parameter estimations in addition to the process infidelity.
```

File: cirq-superstaq/cirq_superstaq/service.py [hunk 7]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+
+        Args:
+            job_id: String corresponding to the CB job id.
+
+        Returns:
+            A string with the job id for the Cycle Benchmarking job created.
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:1049]: looks like this needs to be updated

what is actually returned here? should we be doing any deserialization/parsing/etc?

File: cirq-superstaq/cirq_superstaq/service.py [hunk 8]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+                `cirq.NoiseModel`. Valid strings are "symmetric_depolarize", "phase_flip",
+                "bit_flip" and "asymmetric_depolarize".
+            error_prob: The error probabilities if a string was passed to `noise`.
+                * For "asymmetric_depolarize", `error_prob` will be a three-tuple with the error
+                rates for the X, Y, Z gates in that order. So, a valid argument would be
+                `error_prob = (0.1, 0.1, 0.1)`. Notice that these values must add up to less than
+                or equal to 1.
+                * For the other channels, `error_prob` is one number less than or equal to 1, e.g.,
+                `error_prob = 0.1`.
+        Returns:
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:1011]:
```suggestion

        Returns:
```

File: general-superstaq/general_superstaq/superstaq_client.py [hunk 9]
```diff
@@ -546,6 +546,67 @@ def process_aces(self, job_id: str) -> list[float]:
...
+            serialized_circuits: The serialized process circuits to use in the protocol.
+            num_channels: The number of random Pauli decay channels to approximate error.
+            num_sequences: Number of circuits to generate per depth.
+            depths: Lists of depths representing the depths of Cycle Benchmarking
+                circuits to generate.
+            method: Optional method to use in device submission (e.g. "dry-run").
+            noise: Dictionary representing noise model to simulate the protocol with.
+
+        Returns:
+            A string with the job id for the ACES job created.
```
Comment (richrines1) [inline general-superstaq/general_superstaq/superstaq_client.py:574]:
```suggestion
            A string with the job id for the cycle benchmarking job created.
```

File: cirq-superstaq/cirq_superstaq/service.py [hunk 10]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+        The protocol in detail can be found in: https://arxiv.org/abs/1902.08543.
+
+        Args:
+            target: The device target to characterize.
+            shots: How many shots to use per circuit submitted.
+            circuits: The process circuit to use in the protocol.
+            num_channels: The number of random Pauli decay channels to approximate error.
+            num_sequences: Number of circuits to generate per depth.
+            depths: Lists of depths representing the depths of Cycle Benchmarking
+                circuits to generate.
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:999]: nits (ditto in superstaq_client)
```suggestion
            num_sequences: The number of circuits to generate per depth.
            depths: List of depths of Cycle Benchmarking circuits to generate.
```

File: cirq-superstaq/cirq_superstaq/service.py [hunk 11]
```diff
@@ -973,6 +973,86 @@ def submit_aces(
...
+        method: str | None = None,
+        noise: str | cirq.NoiseModel | None = None,
+        error_prob: float | tuple[float, float, float] | None = None,
+    ) -> str:
+        """Submits the jobs to characterize `target` through the Cycle Benchmarking protocol.
+
+        The protocol in detail can be found in: https://arxiv.org/abs/1902.08543.
+
+        Args:
+            target: The device target to characterize.
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:993]:
```suggestion
            target: The target device to characterize.
```
(ditto in superstaq_client)

**Review by richrines1 — COMMENTED** [review]
should definitely be tested against the server, but afaict this looks good (% some nits)!

eventually we should really make all of the qcvv submission endpoints return specialized jobs, so we could e.g. check their status or use different methods to retrieve/analyze/present different kinds of results

File: cirq-superstaq/cirq_superstaq/service.py (hunk 7 above)
Comment (paaige) [reply cirq-superstaq/cirq_superstaq/service.py]: Yeah, I'll push a subsequent update in a bit.

File: cirq-superstaq/cirq_superstaq/service.py [hunk 12]
```diff
@@ -206,7 +206,8 @@ def get_counts(
         param_resolver: cirq.ParamResolverOrSimilarType = cirq.ParamResolver({}),
         method: str | None = None,
         **kwargs: Any,
-    ) -> dict[str, int]: ...
+    ) -> dict[str, int]:
+        ...
```
Comment (richrines1) [inline cirq-superstaq/cirq_superstaq/service.py:210]: it looks like your local version of `black` is out of date - once you update it and re-run the formatter these changes will be reverted

File: general-superstaq/general_superstaq/superstaq_client.py [hunk 13]
```diff
@@ -546,6 +546,67 @@ def process_aces(self, job_id: str) -> list[float]:
...
+
+    def process_cb(self, job_id: str) -> dict[str, Any]:
+        """Makes a POST request to the "/cb_fetch" endpoint.
+
+        Args:
+            job_id: The job id returned by `submit_cb`.
+
+        Returns:
+            Characterization data including process fidelity
+            and parameter estimates.
```
Comment (richrines1) [inline general-superstaq/general_superstaq/superstaq_client.py:606]:
```suggestion
            Characterization data including process fidelity and parameter estimates.
```

**Review by richrines1 — APPROVED** [review]
will follow up with a post-merge review, but mostly looks good so approving so we can deploy and start testing/demoing!

**Pull request merged**